src/main/java/org/verapdf/wcag/algorithms/semanticalgorithms/consumers/ContrastRatioConsumer.java

```diff
@@ -0,0 +1,212 @@
+package org.verapdf.wcag.algorithms.semanticalgorithms.consumers;
+
+import org.apache.pdfbox.pdmodel.PDDocument;
+import org.apache.pdfbox.pdmodel.PDPage;
+import org.apache.pdfbox.pdmodel.common.PDRectangle;
+import org.apache.pdfbox.rendering.ImageType;
+import org.apache.pdfbox.rendering.PDFRenderer;
+import org.verapdf.wcag.algorithms.entities.INode;
+import org.verapdf.wcag.algorithms.entities.SemanticSpan;
+import org.verapdf.wcag.algorithms.entities.content.TextChunk;
+import org.verapdf.wcag.algorithms.entities.enums.SemanticType;
+import org.verapdf.wcag.algorithms.entities.geometry.BoundingBox;
+
+import javax.imageio.ImageIO;
+import javax.xml.soap.Text;
+import java.awt.*;
+import java.awt.image.BufferedImage;
+import java.io.File;
+import java.io.IOException;
+import java.util.*;
+import java.util.List;
+import java.util.function.Consumer;
+
+public class ContrastRatioConsumer implements Consumer<INode> {
+
+	private Map<INode, Double> colorContrastMap = new HashMap<>();
+	private Map<Integer, BufferedImage> renderedPages = new HashMap<>();
+	private final String sourcePdfPath;
+	//private static final Logger logger = LoggerFactory.getLogger(ContrastRatioConsumer.class);
+	private static final int RENDER_DPI = 144;
+	private static final int PDF_DPI = 72;
+
+	public ContrastRatioConsumer(String sourcePdfPath) {
+		this.sourcePdfPath = sourcePdfPath;
+	}
+
+	public Map<INode, Double> getColorContrastMap() {
+		return colorContrastMap;
+	}
+
+	@Override
+	public void accept(INode node) {
+		List<INode> children = node.getChildren();
+		//check if node is leaf
+		if (!children.isEmpty()) {
+			for (INode childNode : children) {
+				if (childNode.getChildren().isEmpty() &&
+				    SemanticType.SPAN.equals(childNode.getSemanticType())) {
+					calculateContrastRatio((SemanticSpan) childNode);
+				}
+			}
+		}
+	}
```

Comment:
@Override
	public void accept(INode node) {
		if (node.getChildren().isEmpty() && (node instanceof SemanticSpan)) {
			calculateContrastRatio((SemanticSpan) node);
		}
	}

Reply:
ok, refactored it